### PR TITLE
SpMV

### PR DESCRIPTION
**Pull request by kirillgarbar**
## Proposed Changes

SpMV operation added

## Types of changes

What types of changes does your code introduce to GraphBLAS-sharp?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Build and tests pass locally
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/GraphBLAS-sharp.Backend/Vector/SpMV.fs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        =
+        //Until LocalMemSize added to ClDevice as member
+        let error = ref Unchecked.defaultof<ClErrorCode>
+
+        let localMemorySize =
+            Cl
+                .GetDeviceInfo(clContext.ClDevice.Device, OpenCL.Net.DeviceInfo.LocalMemSize, error)
+                .CastTo<int>()
+
+        let localArraySize1 = workGroupSize + 1
```
Comment (gsvgit) [inline src/GraphBLAS-sharp.Backend/Vector/SpMV.fs:24]: Can it be reused in other algorithms to calculate the optimal size of the local array?

File: src/GraphBLAS-sharp.Backend/Vector/SpMV.fs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                        //Reduction
+                        //Check if any part of the row is loaded into local memory on this iteration
+                        if (localPtr.[lid + 1] > blockLowerBound
+                            && localPtr.[lid] < blockLowerBound + workPerIteration) then
+                            let rowStart = max (localPtr.[lid] - blockLowerBound) 0
+
+                            let rowEnd =
+                                min (localPtr.[lid + 1] - blockLowerBound) workPerIteration
+
+                            for jj in rowStart .. rowEnd - 1 do
```
Comment (gsvgit) [inline src/GraphBLAS-sharp.Backend/Vector/SpMV.fs:93]: What is the reason to use such a strange name ```jj```?

File: tests/GraphBLAS-sharp.Tests/BackendCommonTests/SpMVTests.fs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      let byteAdd =
+          Vector.spMV context byteSum byteMul wgSize
+
+      let byteToCOO = Matrix.toCOO context wgSize
+
+      testContext
+      |> correctnessGenericTest 0uy (+) (*) byteAdd (=) q
+      |> testPropertyWithConfig config (getCorrectnessTestName "byte") ]
+
+let tests =
```
Comment (gsvgit) [inline tests/GraphBLAS-sharp.Tests/BackendCommonTests/SpMVTests.fs:121]: Looks like code duplication (look [here](https://github.com/YaccConstructor/GraphBLAS-sharp/pull/48/files/b250012abd937fa40d2e6bc015db74de1461d08c#diff-e13d10256497bf4a38f221c18ec706e53ebafdedde221ef060c4051b8fae87bcR73) or to other tests)

File: src/GraphBLAS-sharp.Backend/Vector/SpMV.fs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                    let mutable sum: 'c option = None
+
+                    while blockLowerBound < workEnd do
+                        let mutable index = blockLowerBound + lid
+
+                        barrierLocal ()
+                        //Loading values to the local memory
+                        for block in 0 .. numberOfBlocksFitting - 1 do
+                            if index < workEnd then
+                                localValues.[lid + block * threadsPerBlock] <- intermediateArray.[index]
```
[the remark this answers is not in the document]
Comment (kirillgarbar) [reply src/GraphBLAS-sharp.Backend/Vector/SpMV.fs:80]: On the last work group `threadsPerBlock` is computed as `numberOfRows - gid + lid`, so `workEnd` will be equal to the last value of row pointers array and indices will not be out of bounds

**Review by artemgl — APPROVED** [review]

**Pull request merged**